### ISSUE TITLE
set_paper_trail_whodunnit warning never goes away

### ISSUE THREAD

**Issue by robmathews** [airblade/paper_trail#645]
Related to https://github.com/airblade/paper_trail/pull/556

I've upgraded papertrail and run across the new requirement to call

    before_filter :set_paper_trail_whodunnit

So far, so good. But when I do this, I continue to get the warning:

    user_for_paper_trail is present, but whodunnit has not been set.
    PaperTrail no longer adds the set_paper_trail_whodunnit
    before_filter for you. Please add this before_filter to your
    ApplicationController to continue recording whodunnit. See the
    PaperTrail readme for an example.

even though I have in fact set the before_filter. I've tried placing the before filter everywhere, but can't seem to get past this warning. I checked the versions table in the do, and whodunnit is being populated.

How do I turn this warning off?

**Comment by jaredbeck** [comment]
Thanks for testing the master branch! The short answer is, you can downgrade from `master` to a released version, like `4.0.0`. But, let's look into it more.

Well, I'm glad the warning was there, because it alerted you to the breaking change, but of course you shouldn't see the warning once you add the callback.  

The logic that controls that warning is pretty simple.

```ruby
      def warn_about_not_setting_whodunnit
        enabled = ::PaperTrail.enabled_for_controller?
        user_present = user_for_paper_trail.present?
        whodunnit_blank = ::PaperTrail.whodunnit.blank?
        if enabled && user_present && whodunnit_blank
          warn # ... etc.
```

Can you tell me more about a particular situation where you see the warning but you think you shouldn't? Which part(s) of that three-part conditional (`enabled && user_present && whodunnit_blank`) is/are `false` for you?

**Comment by robmathews** [comment]
Well then .. that's one answer right there! I forgot we were on master, you merged our pull request eons ago.

**Comment by robmathews** [comment]
(byebug) pp enabled
true
true
(byebug) pp user_present
true
true
(byebug) pp whodunnit_blank
true
true
(byebug)

**Comment by robmathews** [comment]
the application controller has a lot of code, maybe I'm calling papertrail stuff in the wrong order? 

https://gist.github.com/8d8c6d51fcc3a31857fc

**Comment by robmathews** [comment]
Ah .. nope. puts tells me that  set_paper_trail_whodunnit is  called before warn_about_not_setting_whodunnit

when I then debug this code:

    def set_paper_trail_whodunnit
        ::PaperTrail.whodunnit = user_for_paper_trail if ::PaperTrail.enabled_for_controller?
      end

I discover that user_for_paper_trail is nil at this point. However, it is not nil by the time warn_about_not_setting_whodunnit is called.

**Comment by jaredbeck** [comment]
>  set_paper_trail_whodunnit is called before warn_about_not_setting_whodunnit

Yes, the former is a before-callback, the latter an after-callback.

**Comment by jaredbeck** [comment]
It sound like you're on the right track. The order matters. If you're going to use `user_for_paper_trail`, make sure it is `present?` at the time the `set_paper_trail_whodunnit` callback happens.

**Comment by robmathews** [comment]
Ok, so here is the thing - I've got this code:

    class ApplicationController < ActionController::Base
      before_filter :set_paper_trail_whodunnit
      def user_for_paper_trail
          authorized_user_id
      end
      ...

some other subcontroller that needs authorization:

    class FooController < ApplicationController
       before_filter :require_user, :only => [:edit, :update]

Note that require_user is part of Authlogic, it'll set up the authorized_user_id as part of that cycle.

Now, if I move the set_paper_trail_whodunnit down into my subcontroller, it works:

    class FooController < ApplicationController
       before_filter :require_user, :only => [:edit, :update]
       before_filter :set_paper_trail_whodunnit

So the warning is caused because the before_filters in the base controller class are getting called before the before_filters in the subclass.

**Comment by robmathews** [comment]
I wouldn't be closing this - the warning code is broken - I am already calling things in the only order I can.

**Comment by jaredbeck** [comment]
So, in you're app, `require_user` must happen before `set_paper_trail_whodunnit`. How about using `prepend_before_filter :require_user`?

**Comment by robmathews** [comment]
I tried simply putting the call to set_paper_trail_whodunnit as the last line of require_user, it didn't get rid of the warning.

I'm more thinking simply:

skip_filter :warn_about_not_setting_whodunnit

since it has served its purpose.

**Comment by robmathews** [comment]
Yeah, I don't want to edit all my 100s of controllers and change before_filter => prepend_before_filter

**Comment by stereobooster** [comment]
It works the strange way. This code should disable warning in tests, but it is not:

```ruby
RSpec.configure do |config|
  config.before(:each) do
    PaperTrail.enabled = false
    PaperTrail.enabled_for_controller = false
  end
end

```

**Comment by stereobooster** [comment]
Plus it would be very helpful if it would say what was the controller or context. It is very hard to find and very annoying.

**Comment by tilsammans** [comment]
This shouldn't be closed. I've tried all the suggestions, including prepending the before_filter (which I think is a hack) and the warning persists anyway. @jaredbeck

**Comment by jaredbeck** [comment]
Re-opened. Sorry this message is annoying! We have to have a deprecation warning for this breaking change, though. If we don't, people could unexpectedly end up with nulls in the `versions.whodunnit` column, and that would be much worse than an annoying message!

That said, we certainly don't want an unnecessary message! If you're certain that you don't need the deprecation warning, you can, as a workaround, add

```
skip_after_action :warn_about_not_setting_whodunnit
```

I'm open to other suggestions.

**Comment by jaredbeck** [comment]
> Plus it would be very helpful if it would say what was the controller or context.

That sounds like a good idea. I'd be happy to review a PR that adds that feature.

**Comment by jaredbeck** [comment]
Another solution: If you don't want to track `whodunnit`, you can simply override `user_for_paper_trail` to return `nil`.

```
# in application_controller.rb
def user_for_paper_trail
  nil # disable whodunnit tracking
end
```

**Comment by jaredbeck** [comment]
e86c674 improves this warning by adding the suggestions above. Are there any other suggestions before I re-close this issue?

**Comment by istana** [comment]
I appreciate the documentation, I was stuck with the warning for a while. The first link e86c674 doesn't work and it should point to https://github.com/airblade/paper_trail/#4a-finding-out-who-was-responsible-for-a-change. But still...it was a bit confusing until I realized that PaperTrail is enabled for all controllers by default including public ones without *current_user*, but it's enabled only for certain models. Then I used the snippet with "Public user" line and there is no warning anymore.

**Comment by wgrrrr** [comment]
I just ran across this issue while upgrading and have used `skip_after_action :warn_about_not_setting_whodunnit` as a workaround.

I have a hierarchy of base controller classes that handle various situations. It is my intention to handle setting the `whodunit` in one of the subclasses of `ApplicationController`.

**Comment by jaredbeck** [comment]
> The first link [in] e86c674 doesn't work and it should point to https://github.com/airblade/paper_trail/#4a-finding-out-who-was-responsible-for-a-change.

Good catch, thanks. Want to make a PR?

**Comment by jaredbeck** [comment]
> The first link [in] e86c674 doesn't work ...

Fixed by 8f4f408c7b6c84f4fcae4a8330b5e4652a32bd82

> I appreciate the documentation ..

Glad it helped! I'm gonna close this but let me know if you think of anything else.
